Remove email address field from login

kencochrane opened the pull request:
This removes the email prompt when you use docker login, and also removes the ability to register via the docker cli. Docker login, will strictly be used for logging into a registry server.

This picks up where ##16183 left off.

closes: #6400

Not quite what @bfirsh wanted with #18668 but it solves the issue, and keeps backwards compatibility with older version of the engine.

There was one change on engine-api as well: https://github.com/docker/engine-api/pull/105 Not sure what the process is to update the vendored version of that lib.

/cc @bfirsh @dmp42 

Signed-off-by: Ken Cochrane kencochrane@gmail.com

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24028/console):

~~~console

---
ations! "./pkg/..." is safely isolated from internal code.

---> Making bundle: validate-test (in bundles/1.11.0-dev/validate-test)
Congratulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
registry/auth.go:56: unreachable code
exit status 1
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson2539515881580688800.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22050/console):

~~~console

---


Ping statistics for 10.0.0.24:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
ERROR: Failed with exitcode 1 at Sun Feb 21 22:59:06 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson7286523338099341444.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15240/console):

~~~console

---
ations! "./pkg/..." is safely isolated from internal code.

---> Making bundle: validate-test (in bundles/1.11.0-dev/validate-test)
Congratulations! No testing.T found.

---> Making bundle: validate-toml (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
registry/auth.go:56: unreachable code
exit status 1
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson5395704960430134347.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/2160/console):

~~~console

---
 	github.com/docker/docker/autogen/winresources	[no test files]
ok  	github.com/docker/docker/builder	0.072s
ok  	github.com/docker/docker/builder/dockerfile	0.081s
?   	github.com/docker/docker/builder/dockerfile/command	[no test files]
ok  	github.com/docker/docker/builder/dockerfile/parser	0.055s
?   	github.com/docker/docker/builder/dockerfile/parser/dumper	[no test files]
ok  	github.com/docker/docker/builder/dockerignore	0.045s
?   	github.com/docker/docker/cli	[no test files]
--- FAIL: TestOldInvalidsAuth (0.00s)
	config_test.go:143: Should have failed
		Config: &{map[https://index.docker.io/v1/:{joejoe hello   https://index.docker.io/v1/ }] map[]    /tmp/config-test488853038/config.json}
		Got: <nil>
		Expected: Invalid Auth config file
--- FAIL: TestOldValidAuth (0.00s)
	config_test.go:193: Should have save in new form: {
			"auths": {
				"https://index.docker.io/v1/": {
					"auth": "am9lam9lOmhlbGxv",
					"email": ""
				}
			}
		}
--- FAIL: TestOldJson (0.00s)
	config_test.go:262: S
---

~~~

dmp42 commented:
Is this limited to the hub? or will this be true for other registries as well?
Won't this break compatibility for existing registries that would require it?

kencochrane commented:
@dmp42 Are you talking about registering on 3rd party registries?

kencochrane commented:
As you can see the tests are failing, I'll work on getting those fixed. It might be related to the vendored lib change, I mentioned above.

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22051/console):

~~~console

---


Ping statistics for 10.0.0.24:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
ERROR: Failed with exitcode 1 at Sun Feb 21 23:55:51 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson5069521086838961354.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24029/console):

~~~console

---
l (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/registry
registry/auth.go:26: status declared and not used
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson8589770774843127596.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1307/console):

~~~console

---
1T23:57:02Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of cleanup
INFO: Ended at Sun Feb 21 23:57:02 CUT 2016 (1m 30s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\jenkins\hudson2734876546088511348.sh
+ set +e
+ set +x
INFO: Nuking /d/CI
time="2016-02-21T23:57:05Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6509/console):

~~~console

---
sing cache
 ---> ae72d6b72294
Step 48 : ENTRYPOINT hack/dind
 ---> Using cache
 ---> 6b33207d8b36
Step 49 : COPY . /go/src/github.com/docker/docker
 ---> 4a1914747844
Removing intermediate container 901f86d41f99
Successfully built 4a1914747844

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/registry
registry/auth.go:26: status declared and not used
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson8604628513046121603.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15241/console):

~~~console

---
l (in bundles/1.11.0-dev/validate-toml)
Congratulations!  All toml source files changed here have valid syntax.

---> Making bundle: validate-vet (in bundles/1.11.0-dev/validate-vet)
Congratulations!  All Go source files have been vetted.

---> Making bundle: binary (in bundles/1.11.0-dev/binary)
Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev
# github.com/docker/docker/registry
registry/auth.go:26: status declared and not used
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson1227647717817050566.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/2161/console):

~~~console

---
diate container 0cfe7f68f30b
Successfully built 10c4ab0902ba
# WARNING! I don't seem to be running in a Docker container.
# The result of this command might be an incorrect build, and will not be
# officially supported.
#
# Try this instead: make all
#

---> Making bundle: gccgo (in bundles/1.11.0-dev/gccgo)
# github.com/docker/docker/registry
registry/auth.go:26:3: error: 'status' declared and not used
   status        string
   ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson105289745338009634.sh
+ set +x


---

~~~

vdemeester commented:
@kencochrane You can vendor it "manually" until we get it merged in `engine-api`, that way we can review the changes in here in parallel.

For the errors, the `status` variable in not used anymore in your code (here https://github.com/docker/docker/pull/20565/files#diff-0ec52e4c7165f2740e01649dcef5867bR26).

kencochrane commented:
@vdemeester Thanks for the help, I fixed the unit tests, lets see if gordon agrees with me.

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15249/console):

~~~console

---
s
PASS: docker_cli_push_test.go:113: DockerRegistrySuite.TestPushMultipleTags	0.515s
PASS: docker_cli_push_test.go:51: DockerRegistrySuite.TestPushUntagged	0.040s
PASS: docker_cli_by_digest_test.go:159: DockerRegistrySuite.TestRemoveImageByDigest	1.111s
PASS: docker_cli_by_digest_test.go:141: DockerRegistrySuite.TestRunByDigest	1.490s
PASS: docker_cli_by_digest_test.go:208: DockerRegistrySuite.TestTagByDigest	1.167s

----------------------------------------------------------------------
FAIL: docker_cli_v2_only_test.go:71: DockerRegistrySuite.TestV1

[d59907245] waiting for daemon to start
[d59907245] daemon started
docker_cli_v2_only_test.go:116:
    c.Assert(v1Logins, check.Not(check.Equals), 0, check.Commentf("Expected v1 login attempt"))
... obtained int = 0
... expected int = 0
... Expected v1 login attempt

[d59907245] exiting daemon

----------------------------------------------------------------------
PASS: docker_cli_v2_only_test.go:38: DockerRegistrySuite.TestV2Only	0.833s
PASS: docker_cli_pull_local_test.go:202: Do
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24037/console):

~~~console

---
s
PASS: docker_cli_push_test.go:113: DockerRegistrySuite.TestPushMultipleTags	0.540s
PASS: docker_cli_push_test.go:51: DockerRegistrySuite.TestPushUntagged	0.041s
PASS: docker_cli_by_digest_test.go:159: DockerRegistrySuite.TestRemoveImageByDigest	1.295s
PASS: docker_cli_by_digest_test.go:141: DockerRegistrySuite.TestRunByDigest	1.674s
PASS: docker_cli_by_digest_test.go:208: DockerRegistrySuite.TestTagByDigest	1.146s

----------------------------------------------------------------------
FAIL: docker_cli_v2_only_test.go:71: DockerRegistrySuite.TestV1

[d92214045] waiting for daemon to start
[d92214045] daemon started
docker_cli_v2_only_test.go:116:
    c.Assert(v1Logins, check.Not(check.Equals), 0, check.Commentf("Expected v1 login attempt"))
... obtained int = 0
... expected int = 0
... Expected v1 login attempt

[d92214045] exiting daemon

----------------------------------------------------------------------
PASS: docker_cli_v2_only_test.go:38: DockerRegistrySuite.TestV2Only	0.845s
PASS: docker_cli_pull_local_test.go:202: Do
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6517/console):

~~~console

---
+0x131e

goroutine 3422 [select, 80 minutes]:
net/http.(*persistConn).readLoop(0xc8203329a0)
	/usr/local/go/src/net/http/transport.go:976 +0xac7
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:685 +0xc78

goroutine 3423 [select, 80 minutes]:
net/http.(*persistConn).writeLoop(0xc8203329a0)
	/usr/local/go/src/net/http/transport.go:1009 +0x40c
created by net/http.(*Transport).dialConn
	/usr/local/go/src/net/http/transport.go:686 +0xc9d
exit status 2
FAIL	github.com/docker/docker/integration-cli	5400.062s
---> Making bundle: .integration-daemon-stop (in bundles/1.11.0-dev/test-integration-cli)
+++++ cat bundles/1.11.0-dev/test-integration-cli/docker.pid
++++ kill 1899
++++ /etc/init.d/apparmor stop
Clearing AppArmor profiles cache:.
All profile caches have been cleared, but no profiles have been unloaded.
Unloading profiles will leave already running processes permanently
unconfined, which can lead to unexpected situations.

To set a process to complain mode, use the command line tool

---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1321/console):

~~~console

---
	      56	docker-1.11.0-dev.exe.md5
  0%  
100%  
	    New File  		      88	docker-1.11.0-dev.exe.sha256
  0%  
100%  
	    New File  		  27.0 m	docker.exe
  0.0%
  3.6%
  7.3%
 11.0%
 14.7%
 18.4%
 22.1%
 25.8%
 29.5%
 33.2%
 36.9%
 40.6%
 44.3%
 48.0%
 51.7%
 55.4%
 59.1%
 62.8%
 66.5%
 70.2%
 73.9%
 77.6%
 81.3%
 85.0%
 88.7%
 92.4%
 96.1%
 99.8%
100%  

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         1         0         1         0         0         0
   Files :         4         4         0         0         0         0
   Bytes :   54.08 m   54.08 m         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :           1050215296 Bytes/sec.
   Speed :            60093.801 MegaBytes/min.
   Ended : Monday, February 22, 2016 6:13:32 PM

+ ec=0
+ set +x
INFO: Linking the built binary to /d/CI/CI-275e3
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24042/console):

~~~console

---
tainer 15390b3d28d8
Successfully built 32b7c1eb1817

---> Making bundle: validate-dco (in bundles/1.11.0-dev/validate-dco)
These commits do not have a proper 'Signed-off-by:' marker:
 - 829ef430d526d2ebd0473a1e427ab97acb92471f

Please amend each commit to include a properly formatted DCO marker.

Visit the following URL for information about the Docker DCO:
 https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson6566849877889570199.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15255/console):

~~~console

---
XPERIMENTAL is set: building experimental features

---> Making bundle: validate-dco (in bundles/1.11.0-dev/validate-dco)
These commits do not have a proper 'Signed-off-by:' marker:
 - 829ef430d526d2ebd0473a1e427ab97acb92471f

Please amend each commit to include a properly formatted DCO marker.

Visit the following URL for information about the Docker DCO:
 https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson8241602202525404071.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22071/console):

~~~console

---


Ping statistics for 10.0.0.16:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 2ms, Average = 0ms
ERROR: Failed with exitcode 1 at Mon Feb 22 18:20:57 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson660412079742319892.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-vendor [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-vendor/245/console):

~~~console

---
d10
 ---> d1cb2f10b69d
Removing intermediate container 745ae4bd7d10
Step 49 : COPY . /go/src/github.com/docker/docker
 ---> a5c6f7a9a83d
Removing intermediate container 8b16247de02e
Successfully built a5c6f7a9a83d

---> Making bundle: validate-vendor (in bundles/1.11.0-dev/validate-vendor)
The result of ./hack/vendor.sh differs

 M vendor/src/github.com/docker/engine-api/types/auth.go

Please vendor your package with ./hack/vendor.sh.

Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-vendor] $ /bin/sh -xe /tmp/hudson4252042376989914182.sh
+ set +x


---

~~~

calavera commented:
@kencochrane you should be able to vendor engine-api with you fix here. That way we can move this forward.

kencochrane commented:
@calavera Thanks, what is the correct way to vendor engine-api, do I just update the commit hash on hack/vendor.sh or is there more to it then that?

calavera commented:
@kencochrane update the sha and run the script `hack/vendor.sh`, that should vendor the changes. Please, add those changes in a separated commit, so people can review them independently.

You can see an example in https://github.com/docker/docker/pull/19984/commits

stevvooe commented:
LGTM approach wise.

dmp42 commented:
Thanks @kencochrane 

My concerns are:

 * does this affect third-party (non hub) registries and auth servers?
 * do we know if these third-party servers expect a mandatory email address as part of their login endpoint? and if so, what happens with these new clients that don't send them
 * this should really no longer talk/test to `index.docker.io` - all login attempts must go against `registry-1.docker.io` which then redirects to an auth server (usually `auth.docker.io`) <- and this is really what matters nowadays (for the token auth part)

cc @aaronlehmann @dmcgowan @RichardScothern on this

kencochrane commented:
@calavera I bumped up the vendored version of engine-api, thanks for the help.

GordonTheTurtle commented:
Job: Docker-PRs-Win2Lin [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-Win2Lin/22083/console):

~~~console

---


Ping statistics for 10.0.0.20:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 1ms, Average = 0ms
ERROR: Failed with exitcode 1 at Mon Feb 22 23:08:38 CUT 2016.
Build step 'Execute shell' marked build as failure
Performing Post build task...
Could not match :Build timed out  : False
Logical operation result is FALSE
Skipping script  : cd c:\scripts
sh timeout-post-build.sh

END OF POST BUILD TASK 	: 0
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\Jenkins\AppData\Local\Temp\hudson2939500755046082780.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-WoW-TP4 [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-WoW-TP4/1334/console):

~~~console

---
2T23:09:59Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End of cleanup
INFO: Ended at Mon Feb 22 23:09:59 CUT 2016 (1m 30s)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[docker] $ sh -xe C:\Users\jenkins\hudson9083923210415190921.sh
+ set +e
+ set +x
INFO: Nuking /d/CI
time="2016-02-22T23:10:02Z" level=error msg="hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0" 
ERROR:  hcsshim::DestroyLayer - Win32 API call returned error r1=0x80070020 err=The process cannot access the file because it is being used by another process.id=CI flavour=0
INFO: End
---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-userns [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-userns/6538/console):

~~~console

---
 no field or method Len)
runconfig/compare.go:29: a.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:30: b.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:52: a.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:53: b.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-userns] $ /bin/sh -xe /tmp/hudson4854760677856196400.sh
+ set +x


---

~~~

kencochrane commented:
@dmp42

>  does this affect third-party (non hub) registries and auth servers?

Do we have a list of third party registries and auth servers, that we usually check with? I assume Quay, AWS CR, Google CR, and open source registry, anyone else?

> do we know if these third-party servers expect a mandatory email address as part of their login endpoint? and if so, what happens with these new clients that don't send them

Not sure, we should check with them to be sure. How do we normally contact these folks to get their input?

>  this should really no longer talk/test to index.docker.io - all login attempts must go against registry-1.docker.io which then redirects to an auth server (usually auth.docker.io) <- and this is really what matters nowadays (for the token auth part)

I agree, but wasn't sure if that should be part of this PR or a different one. I also wasn't sure what the correct address to use, registry-1.docker.io or auth.docker.io. Hopefully you can help with that one.

GordonTheTurtle commented:
Job: Docker-PRs-gccgo [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-gccgo/2189/console):

~~~console

---
           ^
runconfig/compare.go:30:15: error: reference to undefined field or method 'Slice'
  bCmd := b.Cmd.Slice()
               ^
runconfig/compare.go:52:29: error: reference to undefined field or method 'Slice'
  aEntrypoint := a.Entrypoint.Slice()
                             ^
runconfig/compare.go:53:29: error: reference to undefined field or method 'Slice'
  bEntrypoint := b.Entrypoint.Slice()
                             ^
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-gccgo] $ /bin/sh -xe /tmp/hudson8983969937823472554.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs/24057/console):

~~~console

---
Len)
runconfig/compare.go:29: a.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:30: b.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:52: a.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:53: b.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
Build step 'Execute shell' marked build as failure
Archiving artifacts
[PostBuildScript] - Execution post build scripts.
[Docker-PRs] $ /bin/sh -xe /tmp/hudson6220433426173313993.sh
+ set +x


---

~~~

GordonTheTurtle commented:
Job: Docker-PRs-experimental [FAILED](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/15270/console):

~~~console

---
 no field or method Len)
runconfig/compare.go:29: a.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:30: b.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:52: a.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
runconfig/compare.go:53: b.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
Build step 'Execute shell' marked build as failure
[PostBuildScript] - Execution post build scripts.
[Docker-PRs-experimental] $ /bin/sh -xe /tmp/hudson6038605884557696349.sh
+ set +x


---

~~~

BrianBland commented:
@kencochrane @dmp42 as far as I can tell, we've only ever sent the email address to the POST (create account) endpoint, and never to the GET (login) endpoint, so this shouldn't break any existing auth server implementations

BrianBland commented on registry/auth.go:
```diff
@@ -39,93 +35,34 @@ func loginV1(authConfig *types.AuthConfig, registryEndpoint *Endpoint) (string,
 	}
-
-	if respStatusCode == 201 {
+	if resp.StatusCode == 200 {
```
Nit: replace these status code numbers with `http.StatusOK`, `http.StatusUnauthorized`, and `http.StatusForbidden`

dmp42 commented:
@BrianBland fair - thanks for clarifying.

@kencochrane I guess my last concern will be about compatibility of that new engine with the old credentials file. Maybe we should have a test that for that?

calavera commented:
> Maybe we should have a test for that?

adding a test for backwards compatibility would be lovely.

kencochrane commented:
@BrianBland thanks, I made the change.

@dmp42 @calavera there are a few tests for that already, I have put some links to the code, so it is easier to find.
- https://github.com/kencochrane/docker/blob/remove_email_on_login/cliconfig/config_test.go#L146-L194
- https://github.com/kencochrane/docker/blob/remove_email_on_login/cliconfig/config_test.go#L222-L265
- https://github.com/kencochrane/docker/blob/remove_email_on_login/cliconfig/config_test.go#L222-L265

I also added a new one here:
- https://github.com/kencochrane/docker/blob/remove_email_on_login/cliconfig/config_test.go#L222-L265

Does that work, or would you like another one, covering something else?

Also, I noticed that when I bumped the vendored version of engine-api, it pulled in some other changes that are conflicting with my PR. It looks like they came from this PR: https://github.com/docker/engine-api/pull/97

Is it my responsibility to fix the code that is broken when the vendor version was bumped, or are those changes happening elsewhere?

calavera commented:
@kencochrane that works, but it looks like this doesn't compile:

```
20:07:27 # github.com/docker/docker/image/v1
20:07:27 image/v1/imagev1.go:34: v1Image.ContainerConfig.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
20:07:30 # github.com/docker/docker/runconfig/opts
20:07:30 runconfig/opts/parse.go:236: cannot use strslice.New(parsedArgs[1:]...) (type strslice.StrSlice) as type *strslice.StrSlice in assignment
20:07:30 runconfig/opts/parse.go:239: cannot use strslice.New(*flEntrypoint) (type strslice.StrSlice) as type *strslice.StrSlice in assignment
20:07:30 runconfig/opts/parse.go:371: cannot use runCmd (type *strslice.StrSlice) as type strslice.StrSlice in field value
20:07:30 runconfig/opts/parse.go:375: cannot use entrypoint (type *strslice.StrSlice) as type strslice.StrSlice in field value
20:07:34 # github.com/docker/docker/runconfig
20:07:34 runconfig/compare.go:20: a.Cmd.Len undefined (type strslice.StrSlice has no field or method Len)
20:07:34 runconfig/compare.go:20: b.Cmd.Len undefined (type strslice.StrSlice has no field or method Len)
20:07:34 runconfig/compare.go:24: a.Entrypoint.Len undefined (type strslice.StrSlice has no field or method Len)
20:07:34 runconfig/compare.go:24: b.Entrypoint.Len undefined (type strslice.StrSlice has no field or method Len)
20:07:34 runconfig/compare.go:29: a.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
20:07:34 runconfig/compare.go:30: b.Cmd.Slice undefined (type strslice.StrSlice has no field or method Slice)
20:07:34 runconfig/compare.go:52: a.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
20:07:34 runconfig/compare.go:53: b.Entrypoint.Slice undefined (type strslice.StrSlice has no field or method Slice)
20:08:18 Unable to find image 'busybox:latest' locally
```

TIP: Use `git rebase --interactive master` to put the vendor commit first and squash the other two together, so you can append changes to the last commit :smile_cat:

stevvooe commented:
@kencochrane @calavera I can submit a patch to docker/docker to clean up the changes from docker/engine-api#97 if that will make things easier.

kencochrane commented:
@stevvooe that would be great, I just rebased from master, and ran the tests locally, and those tests are still failing for me.

thaJeztah commented on api/client/login.go:
```diff
@@ -26,7 +26,6 @@ func (cli *DockerCli) CmdLogin(args ...string) error {
 
 	flUser := cmd.String([]string{"u", "-username"}, "", "Username")
 	flPassword := cmd.String([]string{"p", "-password"}, "", "Password")
-	flEmail := cmd.String([]string{"e", "-email"}, "", "Email")
```
Wonder if we should *remove* the flag, or only mark it deprecated for now (and simply ignore its value). Our deprecation policy is to mark options deprecated, but keep them around for two more releases, see for example: https://github.com/docker/docker/blob/master/docs/deprecated.md#hostconfig-at-api-container-start

kencochrane replied:
@thaJeztah if we mark it as deprecated, does that mean, we keep it there, we just ignore the value that is passed in? I'm fine with what ever you think is best, I just want to make sure I understand what you are saying, so that I can implement it correctly.

thaJeztah replied:
@kencochrane yes, that's what I was thinking; reason for keeping the flag is that we may break scripts that people use, and they will end up with:

```
flag provided but not defined: --email
See 'docker run --help'.
```

I think it's better to keep the flag for that reason

kencochrane replied:
@thaJeztah OK, I'll add it back. Should we print a warning message to the screen when they use it? Also, were in the docs do we list the deprecated features, so people know?

thaJeztah replied:
@kencochrane If you add it back, and prefix the flag with a `#`, then the `flag` package should automatically output a deprecation message. See for example; https://github.com/docker/docker/blob/131e2bf12b2e1b3ee31b628a501f96bbb901f479/api/client/ps.go#L28

And, yes, it should be added to the documentation (in the `deprecated.md` file I linked above)

kencochrane replied:
@thaJeztah ok, I did as you asked, I added the flag back in, and added a note in deprecated.md let me know how that looks.

kencochrane commented:
@bfirsh I added the messages you requested, this is what it looks like (see below), let me know what you think.

```
$ docker login
Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one.
Username: myusername
Password: badpassword
Error response from daemon: Wrong login/password, please try again. Haven't got a Docker ID? Create one at https://hub.docker.com
```

bfirsh commented:
@kencochrane Looks great, thanks Ken!

kencochrane commented on api/client/login.go:
```diff
@@ -26,7 +26,9 @@ func (cli *DockerCli) CmdLogin(args ...string) error {
-	flEmail := cmd.String([]string{"e", "-email"}, "", "Email")
+
+	// Deprecated in 1.11: Should be removed in docker 1.13
+	cmd.String([]string{"#e", "#-email"}, "", "Email")
```
@thaJeztah here is where the flag was added back, and the #'s added for deprecation

thaJeztah replied:
Yup, that's what I had in mind, thanks!

thaJeztah commented on cliconfig/config_test.go:
```diff
@@ -164,7 +161,7 @@ func TestOldValidAuth(t *testing.T) {
 	js := `username = am9lam9lOmhlbGxv
-email = user@example.com`
+	email = user@example.com`
 	if err := ioutil.WriteFile(fn, []byte(js), 0600); err != nil {
```
Not sure this whitespace should be removed? It's part of the string

aaronlehmann commented on api/client/login.go:
```diff
@@ -85,6 +89,11 @@ func (cli *DockerCli) configureAuth(flUser, flPassword, flEmail, serverAddress s
+		if isDefaultRegistry {
+			// if this is a defauly registry (docker hub), then display the following message.
+			msg := "Login with your Docker ID to push and pull images from Docker Hub. If you don't have a Docker ID, head over to https://hub.docker.com to create one."
+			fmt.Fprintf(cli.out, "%s\n", msg)
```
nitpick: `fmt.Fprintln(cli.out, msg)` is ismpler.

aaronlehmann commented on vendor/src/github.com/docker/engine-api/types/auth.go:
```diff
@@ -5,7 +5,6 @@ type AuthConfig struct {
 	Username      string `json:"username,omitempty"`
 	Password      string `json:"password,omitempty"`
 	Auth          string `json:"auth,omitempty"`
-	Email         string `json:"email"`
```
I'm not sure it's a good idea to remove this field from the struct. It means that when docker saves the config file, it will strip out email addresses that were already present. Then it will no longer be possible to use that config file with an earlier version of the client, even if you don't touch any of the entries in question.

Maybe this is not a big concern to others. Thoughts?

dmcgowan replied:
Will test this, I don't see why it should be necessary to keep around.

dmcgowan replied:
Testing with 1.9.1 after login with this change

```
$ docker login
Username (dmcgowan): 
Error response from daemon: Registration: "Wrong email format (it has to match \"[^@]+@[^@]+\\.[^@]+\")"
```

dmcgowan replied:
Will test it with omitempty to see if that is more backwards compatible.

dmcgowan replied:
With further testing, if you login with change, then you can still use the config file to login to other servers and push from an older client, just an attempt to login again will fail. With the omitempty flag added, the email will not be added when logging in with this change so downgrade login will still fail, but if you previously logged in with older client, logging in with this change will not strip the email. I don't think it will have a hugely negative impact to completely remove but still think omitempty is probably the correct solution. WDYT?

kencochrane replied:
@dmcgowan I think you are right the omitempty flag is probably a better approach to support backwards compatibility.

aaronlehmann commented on registry/auth.go:
```diff
@@ -39,93 +35,37 @@ func loginV1(authConfig *types.AuthConfig, registryEndpoint *Endpoint) (string,
+		return "", fmt.Errorf("Wrong login/password, please try again")
+	} else if resp.StatusCode == http.StatusForbidden {
+		if loginAgainstOfficialIndex {
+			return "", fmt.Errorf("Login: Account is not Active. Please check your e-mail for a confirmation link.")
```
Unrelated to this PR, but `Active` should not be capitalized.

aaronlehmann commented on registry/auth.go:
```diff
@@ -39,93 +35,37 @@ func loginV1(authConfig *types.AuthConfig, registryEndpoint *Endpoint) (string,
+			return "", fmt.Errorf("Login: Account is not Active. Please check your e-mail for a confirmation link.")
 		}
+		// *TODO: Use registry configuration to determine what this says, if anything?
+		return "", fmt.Errorf("Login: Account is not Active. Please see the documentation of the registry %s for instructions how to activate it.", serverAddress)
```
Same here (capitalization of Active).

aaronlehmann commented on registry/auth.go:
```diff
@@ -39,93 +35,37 @@ func loginV1(authConfig *types.AuthConfig, registryEndpoint *Endpoint) (string,
-	if respStatusCode == 201 {
+	if resp.StatusCode == http.StatusOK {
+		return "Login Succeeded", nil
+	} else if resp.StatusCode == http.StatusUnauthorized {
```
I feel this series of if/else statements which check the status code would be slightly easier to read as a switch statement, but this is cosmetic.

dmcgowan replied:
Not sure I agree with this, mostly because if/else seems to used more for status code checking throughout the code. Whenever I use a switch statement it inevitably gets changed back whenever the error checking becomes more sophisticated.

calavera commented:
@kencochrane, @dmcgowan I just merged docker/engine-api#115. Feel free to vendor it here.

thaJeztah commented:
Should we split the vendor and slice changes to a separate PR? Looks like that change is now bigger than the actual change for removing email

aaronlehmann commented:
@thaJeztah: Doing so would require creating a special branch of engine-api with only a subset of its recent changes. engine-api is now out of sync with docker/docker because of the email change and the strslice changes. It's a bit troubling for it get so far out of sync that any PR needing to bring in a new version becomes so messy.

calavera commented:
The vendoring of engine-api just landed in master and it's not necessary here anymore. That will make merging these changes much more simple after rebasing. Go @kencochrane go! :metal:

dmcgowan commented:
Looks like adding the `omitempty` flag made it vendor-able on its own,  will rebase it out :simple_smile:

kencochrane commented:
@dmcgowan let me know if you need my help with anything.

icecrime commented:
Related failure on janky?

```
01:20:53 FAIL: docker_cli_pull_local_test.go:365: DockerRegistryAuthSuite.TestPullWithExternalAuth
01:20:53 
01:20:53 docker_cli_pull_local_test.go:393:
01:20:53     c.Assert(string(b), checker.Contains, "email")
01:20:53 ... obtained string = "" +
01:20:53 ...     "{\n" +
01:20:53 ...     "\t\"auths\": {\n" +
01:20:53 ...     "\t\t\"127.0.0.1:5000\": {}\n" +
01:20:53 ...     "\t},\n" +
01:20:53 ...     "\t\"credsStore\": \"shell-test\"\n" +
01:20:53 ...     "}"
01:20:53 ... substring string = "email"
```

dmcgowan commented:
Yeah, failed after rebase, fixing now

kencochrane committed:
Remove email address field from login

This removes the email prompt when you use docker login, and also removes the ability to register via the docker cli. Docker login, will strictly be used for logging into a registry server.

Signed-off-by: Ken Cochrane <kencochrane@gmail.com>

vdemeester commented:
code LGTM 🐯

aaronlehmann commented:
LGTM

thaJeztah commented:
@aaronlehmann you skipped "docs-review". I had a quick look, and think it's ok, just thought I'd should mention :innocent:

I'll add the impact labels as well

kencochrane commented:
woohoo! Thank you @vdemeester @dmcgowan @calavera @thaJeztah @aaronlehmann @stevvooe for all of the help getting this completed.

stevvooe commented:
:clap:

vdemeester commented:
:clap: :clap:

stevvooe commented:
> :clap: :clap:
:bulb:

bfirsh commented:
Yay! Thanks @kencochrane! :cake:

RichardScothern commented:
:clap: :clap: :smile_cat:

agherzan commented:
This broke login on 1.10 for example using a new client even if `DOCKER_API_VERSION` is provided. Should this be a separate issue?

thaJeztah commented:
The `DOCKER_API_VERSION` env-variable was designed for debugging purposes, and does not guarantee full backwards compatibility with older API versions (it's just an "override" to _force_ the client to use a `v1.x` API endpoint). Docker 1.13 introduces API version negotiation, and _does_ take older API versions into account, but the oldest API version negotiated will be API 1.24 (which corresponds with docker 1.12)

agherzan commented:
Thank you. That clarifies it.
